### ISSUE TITLE
The entire use of NgModule should be deprecated

### ISSUE THREAD

**Issue by ywarezk** [angular/angular#37904]
The aim of this post is to create a discussion on the topic, there is also a really simple solution I offered, but I really like to hear your thoughts... 

It is my opinion that creating a module system in Angular **NgModule** in addition to the already existing  module system of Javascript is a bad idea and should be changed, or at least the developers should have a choice to work in a non NgModule mode...

Here is why I think NgModules will hold angular progress back: 
1. Bundle size - if I'm creating a library or installing a library containing a module with 10 components, I want to use one of those components in my app, but I only want to use one of them, I add the module to my imports array and my bundle will contain 9 components I'm not even using - this is incorrect with the new Ivy. 
2. Maybe this is why angular material wraps almost every component in the library in it's own module, a bit ugly in my opinion but necessary because of the NgModule system, still it requires me adding tons of stuff to my imports library. 
3. I know how to solve 2 I will just create CommonModule with the imports in one module and then every module that use them will add that module to the imports, but isn't that pattern messing my bundle size as well? 
For example let's say I have common stuff that every module is using and than a bunch of lazy loaded module that each we can divide to a group that use the common ones and another group of other common modules, I don't want to add everything to to CommonModule so I would have do to Common1Module, Common2Module, etc. for every group cause if I throw everything on the CommonModule it will effect the bundle size of every lazy loaded part of my app won't it? 
4. It's not a reason but it's a big clue that the NgModule system is not needed, there is no module system in any other frontend framework or library that i'm aware of, including React and Vue, they simply use the module system in JS. 
5. Learning curve - While it's pretty easy for people to learn that they need modules, separation of concerns is a pretty well understood concept for people and they know they should split their code to "libraries", so the angular module system is pretty simple to understand, but usually when I teach and talk to angular developers they get lost how the DI works cause the module system heavily intertwined in how the DI is searching for providers.
6. HttpClientModule is a good example here, let's say I'm creating a library which contain an NgModule. Obviously that module can be added to the imports of a lazy loaded module, I cannot add HttpClientModule to my library cause it can create another instance of HttpClient if it is used, and then I can mess up my interceptor. So now I have to worry about peerDependencies as well. Example ngrx libraries require us to add HttpClientModule. We already have a peerDependency mechanism inside javascript modules, now we have to do the same with angular

Currently I see the Ivy renderer allows us to create components without modules, but it is still heavily limited in what the component can use in terms of other directives and providers, without a module that is wrapped and supplies them. 

It somewhat seemed to me that the NgModules system is something that was taken from the Angular.js ecosystem and just moved along to the current Angular and I don't think it's needed at all, and not only that it creates more harm than good.

I think the NgModule requires a drastic change which will have to cause a drastic change in the DI, and we will not achieve the bundle size of the competitors React and Vue unless a major change here is achieved . 
I think NgModule has more cons than pros and would love to hear your thoughts.

Making a really general proposal from this post these are the changes I think should be made:
1. The DI system should not automatically create different trees for modules and for the view, rather it should be a simple solution that should be inspired from the library TypeDI. 
You import the DI and you register services to it, and if you want to create a tree structure in your DI you do it manually, you import the DI system and fork it so a certain component can have it's own dependencies. 
Or at least to support backward compatibility allow the DI to work in manual mode, just like we did with the change detection strategy. The manual mode of the DI will enable the developer to control the forking of the di to trees and thus he can choose to work in a non NgModule mode.
2. After that remove the modules entirely, every component that use a certain directive component should use the js module system and use a certain import.
Maybe this step can also somehow be integrated in the compiler in some way so based on the selector angular will know which component to integrate without using the js module system. I'm not entirely strong on the compilation process so I don't think I'm qualified to make a suggestion here.

**Comment by mlc-mlapis** [comment]
[**@ywarezk**](https://github.com/ywarezk) I am afraid that you really put only questions on the table and no answers or proposals, as you said. All of the points above were discussed many times in many places and probably you are looking form an analytical document that summarizes all pros and cons and the background of all those reasons.

**Comment by ywarezk** [comment]
> [**@ywarezk**](https://github.com/ywarezk) I am afraid that you really put only questions on the table and no answers or proposals, as you said. All of the points above were discussed many times in many places and probably you are looking for an analytical document that summarises all pros and cons and the background of all those reasons.

You are right, I updated with my 50 cents suggestion. 
I think An actual discussion containing suggestions from other angular developers is really needed and should be organised in a thread. I really think this issue will be a major set back in angular moving forward.

**Comment by mlc-mlapis** [comment]
**@**[**ywarezk**](https://github.com/ywarezk) I can imagine a particular addition to the actual DI, when you would be able to inject manually more things, like providers or view providers are today on a component metadata level to allow more independent standalone components. But the complete replacement of the DI system is probably outside of reality, and I would vote actively against that idea.

> The DI system should not automatically create different trees for modules and for the view, rather it should be a simple solution that should be inspired from the library TypeDI.  
> You import the DI and you register services to it, and if you want to create a tree structure in your DI you do it manually, you import the DI system and fork it so a certain component can have it's own dependencies.  
> Changing the DI system has to be first before changing the module system.

**Comment by ywarezk** [comment]
> **@**[**ywarezk**](https://github.com/ywarezk) I can imagine a particular addition to the actual DI, when you would be able to inject manually more things, like providers or view providers are today on a component metadata level to allow more independent standalone components. But the complete replacement of the DI system is probably outside of reality, and I would vote actively against that idea.
> 

Yes maybe you are right, maybe not a replacement is needed, rather than an addition. 
Allow developers to work with NgModules the old fashion way, with the DI on automatic mode, but also allow developers to choose to abandon the NgModules, and in that mode of work they will have to manually take care of the DI.

This way I guess you are not breaking backward compatibility, but also allowing developers to take the framework to the next level

Angular did something similar to approach with the framework's change detection. Angular ships with automatic change detection but knowing it has a downside it allows me to choose a different strategy. 
Same concept should be applied with the DI and the NgModules. 
Support NgModule and the old DI but let developers choose to abandon them with a different suitable strategy just like angular did with change detection.

**Comment by matthewjh** [comment]
@ywarezk, as far as I know, number 1 is not true. If you look in the code generated by the Angular compiler, you should see that there are no NgModule declarations arrays present. In the compiled code, components and directives are referenced directly by the components and directives that use them. These references are not created for components and directives that are not actually used. Angular Ivy does pretty much exactly what you said:

> Maybe this step can also somehow be integrated in the compiler in some way so based on the selector angular will know which component to integrate without using the js module system. 

Therefore, if you take a vanilla Angular CLI project, you should be able to prove that unused components, directives, and pipes are removed from production bundles.

Where NgModules _do_ have a problem with tree shaking is with providers, which is why tree-shakable providers were introduced.

Having said all this, I do think the Angular team needs to take a long-term look at NgModules, DI, and service provision mechanisms and come up with a plan to optimises usability and performance. I think there are some benefits to the NgModule concept. As far as I am aware, AngularDart did _not_ have NgModules -- instead resorting to provider arrays and declaring directives and pipes directly on components -- but they planned to add it (for a reason, I assume).

The idea of an NgModule as a bundle of DI providers + components + directives + pipes is quite nice, but in my view, it has become far less cohesive over time with the introduction of tree-shakable providers. It makes no sense for people to be told to use NgModules in the main only for components, directives, and pipes (because these can be tree shaken), but not for services (because they can't be tree shaken). NgModules best makes sense as a ubiquitous concept and if they are used consistently. Now that services are mostly out of the NgModule picture, the concept starts to fall apart. Today it feels awkward. This is why I feel there needs to be a plan rather than a jumble of uncoordinated changes. 

The only sensible alternative to NgModules I am aware of is what AngularDart does:
1. Reference consumed directives and pipes in @ Component decorator
2. Initialise Injectors with provider arrays
3. Use provider arrays to push around sets of providers
4. Provide providers through components and component injectors

```
import 'package:angular/angular.dart';
import 'package:angular/experimental.dart';
import 'package:angular/security.dart';

@Component(
  selector: 'comment',
  templateUrl: 'comment_component.html',
  styleUrls: ['comment_component.css'],
  directives: [
    CommentComponent,
    NgForIdentity,
    NgIf,
  ],
  changeDetection: ChangeDetectionStrategy.OnPush,
)
class CommentComponent {
  final DomSanitizationService _sanitizer;

  @Input()
  Map comment;

  bool hidden = false;

  CommentComponent(this._sanitizer);

  SafeHtml get content =>
      _sanitizer.bypassSecurityTrustHtml(comment['content']);

  String get showCommentText => '+${comment['comments_count']}';

  void toggleVisibility() {
    hidden = !hidden;
  }
}

```

However, like I said, even AngularDart planned to introduce an NgModule analogue:
https://github.com/dart-lang/angular/pull/1032

I'm not sure whether this went ahead or not or what the reasoning behind it was but maybe @matanlurey can comment.

At the end of the day, there's no point in getting rid of NgModules for the sake of it. What _does_ need to happen is real thought about the present mixture of NgModules, Component-level provision, services, tree-shakable providers, and an analysis of issues including:
1. Boilerplate reduction
2. Dead code elimination
3. Consistency and ubiquity
4. Testability
I'm sure something very good can come out of such a process if there is the will.

**Comment by ywarezk** [comment]
> @ywarezk, as far as I know, number 1 is not true. If you look in the code generated by the Angular compiler, you should see that there are no NgModule declarations arrays present. In the compiled code, components and directives are referenced directly by the components and directives that use them. These references are not created for components and directives that are not actually used. Angular Ivy does pretty much exactly what you said:
> Therefore, if you take a vanilla Angular CLI project, you should be able to prove that unused components, directives, and pipes are removed from production bundles.

Amazing!
Thanks for the correction, after checking you are absolutely right .
Corrected the post
thanks :)

**Comment by mlc-mlapis** [comment]
[**@ywarezk**](https://github.com/ywarezk) Yep, but it also means that one component from the outside point of view doesn't mean that inner complexity is trivial. It only means that there is only one exported.

**Comment by ywarezk** [comment]
> [**@ywarezk**](https://github.com/ywarezk) Yep, but it also true that one component from the outside point of view doesn't mean that inner complexity is trivial. It only means that there is only one exported.

Yes of course
If I use a certain component from a module, that component will be added along to my code along with all the other complexities and inner components, directives, pipes, services that this component is using. 
But that will also be the case now when angular using multiple NgModules for angular material and will also be the case if angular will create a single module for the entire material library. Am I wrong here? After Ivy can't we create a single module for Material without the sideeffect of increasing the bundle size more that needed? And if im right shouldn't it be one of the goals for the next version?

**Comment by mlc-mlapis** [comment]
[**@ywarezk**](https://github.com/ywarezk) It depends. Multi-point entries to packages always tend to be always less flexible and less optimized from the tree-shaking point of view. It's possible, of course, but the design should be the right one. Also adding a new NgModule to the game represents usually 5 lines of code so it's probably nothing from the bundle size point of view.

**Comment by spock123** [comment]
I really like the modules system.
It keeps me from having to import a gazillion components, pipes in every component. What a mess.

I like the way that modules are lazy loadable in a very easy way, and that you can break them down to one module per component if you really want to (but don't need to).

I don't see many advantages ditching DI and ngModules at all.

I come from React and it's one big mess.. it's horrible to work with in comparison.

**Comment by sod** [comment]
It's the burden of having a string based template engine. Someone has to tell the compiler, what the heck `<my-component my-direcitve="my-variable | my-pipe">` actually means. And you can't have ES imports in a html file.

I wonder if a config file, that defines some globs where to look for components/directives/pipes, so at least angular can automatically resolve those tokens by default. Similar to how symfony autowire works.

**Comment by spock123** [comment]
@sod I'm worried that by introducing more "magic imports" it will be more difficult to maintain code. If you auto-import components like that,  you effectively create global namespacing, and that is bad.

I really like that I know that components and directives/pipes I create in a module, stay in that module unless I export them.

Effectively, ngModules kinda serve like what you want, you import the dependencies you need, and magically all module components will gain access to those. To me it sounds like you are advocating putting all components in the main application module. I might have misunderstood you, but that's the consequence I get from it.

**Comment by Airblader** [comment]
Such a config file would break locality again.

**Comment by Javaman2** [comment]
[Agreed](https://github.com/angular/angular/issues/38056) since the Javascript adoption of the their way of using the module system, we don't need the ngModule system in it's current form any longer.    

- Advantages of the index.js system is that we are easily able to create multi import statements.

This alone is the best reason. 

But due to legacy issues the NgModule system can still live on.  Just allow us an option to point to an index.js file instead of a Ngmodule file.  All the arguments against this concept listed above are simply not true.  Why?  Because if we create a Angular library, all ngModules collapse to the index.js system anyway!  This means that there is nothing that won't work.  

Just give us an option to point to an index.js file and we're done.  Ngmodule system lives, and for those of us wanting to move away from it, we're satisfied too.

**Comment by yjaaidi** [comment]
No rush. It's coming.
Cf. https://angular.io/guide/roadmap#simplified-angular-mental-model-with-optional-ngmodules

**Comment by LayZeeDK** [comment]
@yjaaidi I hope so, but:
1. That's what they said more than 2 years ago.
2. It's still in the "Future" section which means that it's not actively being worked on and might not happen.

**Comment by LayZeeDK** [comment]
That being said, I wrote my early thoughts on this issue in "[Angular Revisited: Tree-shakable components and optional NgModules](https://dev.to/this-is-angular/angular-revisited-tree-shakable-components-and-optional-ngmodules-36d2)".

**Comment by ifeltsweet** [comment]
Having spent many years with Angular (and loving most of it), I've recently used Vue on a medium-sized project and then React on another similar project. Let me tell you what a joy it was not having to deal with NgModules and just relying on the ES6 modules that are natively supported by most tooling.

My productivity went up and the dread of creating many small, isolated components completely went away in turn having a positive effect on the codebase. I really think the NgModules hinder or at least slow-down composability.

> I really like the modules system.
> It keeps me from having to import a gazillion components, pipes in every component. What a mess.

@spock123 You can easily do that with native JavaScript modules as well. You can even leverage tree-shaking if you don't need all components from a shared module.

**Comment by matthewjh** [comment]
@ifeltsweet the difference is that with ngmodules, you can let the angular compiler map from the selectors you are using in a component -> the directive and component implementations in the transitive ngmodule scope. This may not be a big deal with components because the mapping is clear, but it's more complicated for directives esp. if you use funky selectors to add custom behaviours.

**Comment by ywarezk** [comment]
Although the optional `NgModule` is on the roadmap, it will probably take a very long time until it's implemented.
Until then we can add mini features, that will reduce the usage of `NgModule`.

I opened this issue as well: [issue #39695](https://github.com/angular/angular/issues/39695)

TLDR;
add **declarations** array in the component metadata.

I think it's really easy to implement and it should be pushed forward as one of the steps on the way of removing the NgModules.

**Comment by magicmatatjahu** [comment]
Maybe context like in React will resolve the problem? If Angular will treat modules also like components, then we can do something like:

```ts
@NgModule({
  selector: "some-module",
  providers: [SomeProvider1, SomeProvider2],
})
class SomeModule {}
```

```html
<some-module>
  <some-component>
  </some-component>
</some-module>
```

and then `some-component` component will search providers by context: first in itself metadata (viewProviders), then in `some-module` module. Also, by this approach Angular will be able to create lazy-modules without using router:

```html
<app-module>
  <some-component>
    ...body
  </some-component>
  <div *ngIf="{condition}">
    <lazy-module>
      ...components
    <lazy-module>
  </div>
</app-module>
```

This solution has some cons and pros, but it'll be huge breaking change in Angular ecosystem.

**Comment by LayZeeDK** [comment]
@magicmatatjahu components already work like that. This is known as component-level provider scope and the element injector. There was some work to do a more advanced node injector a few years ago, but I don't think it ever got done.

**Comment by magicmatatjahu** [comment]
@LayZeeDK Thanks for response! Yeah, I know about `NodeInjector`, but I had in mind that this syntax:

```html
<app-module>
  <some-component>
    ...body
  </some-component>
</app-module>
```

is actually this in current implementation:

```ts
@NgModule({
  declarations: [SomeComponent], // with "some-component" selector
  providers: [SomeProvider1, SomeProvider2],
})
class SomeModule {}
```

generally there is also a router etc..., but I hope you know what I mean :)

However React using JSX treats component as normal JS function and then perform action by reference to this function, I mean render component, update it etc... Angular performs node-tree using html, so it uses strings, and we can have a duplicated selectors of components/directives/pipes between module declaration. Next problem is that (related to strings as selector of nodes), making reference to appropriate selector: `<some-component>` -> `SomeComponent` is probably impossible without declaration of reference to this component inside `NgModule` metadata... and I realized that this solution won't work even when using `providedIn` in single declaration definition 😄

**Comment by LayZeeDK** [comment]
Do you mean so that `<app-module>` doesn't get rendered in the DOM?

**Comment by magicmatatjahu** [comment]
`<app-module>` should be renderer (based on module declaration), but in my idea is a problem, how to tell to Angular that definition (factory) of `<some-component>` is `SomeComponent` (definition in TS code) component.

**Comment by LayZeeDK** [comment]
My mistake, I meant to ask whether you thought that `<some-module>` shouldn't be rendered.

**Comment by Maximaximum** [comment]
Heads up! There's an Angular issue open that builds on the idea of optional NgModules and suggests to make the declarable's selectors overridable and configurable, so that people can use them just the way we are currently using providers.

If you'd like to be able to use optional NgModules and the extended capabilities, please feel free to vote for the issue: https://github.com/angular/angular/issues/35646

If the issue gets 20+ thumbups until June 23, 2021, it will be considered by the Angular team for implementation.

**Comment by misha98857** [comment]
Now standalone components are a stable part of Angular v15.

**Comment by mlc-mlapis** [comment]
We hope both ways will be available because NgModules still represent significant advantages when developing and maintaining large projects.

**Comment by MattiaMalandrone** [comment]
> We hope both ways will be available because NgModules still represent significant advantages when developing and maintaining large projects.

This is exactly the motivation behind my question!

**Comment by magicmatatjahu** [comment]
If you are writing a library/shareable logic with e.g. services, helpers etc then modules are great for that. Similarly, when you divide applications into domains, a module makes more sense in my opinion than a standalone component (which can also be used as a domain, but a module is better imho).

**Comment by spock123** [comment]
Modules are invaluable for very large projects,  I don't see them ever going away. I certainly hope they won't.

**Comment by hannah23280** [comment]
Agree to those who believe Modules are useful, especially large projects. The only question now is whether the angular engineer will have headache having to maintain two choices in future, fi they decide to keep both

**Comment by pfeileon** [comment]
> Modules are invaluable for very large projects, I don't see them ever going away. I certainly hope they won't.

What do they offer but a verbose way to wrap stuff in a class object? How is that any different from wrapping stuff in an object or array literal?

**Comment by DaSchTour** [comment]
> > Modules are invaluable for very large projects, I don't see them ever going away. I certainly hope they won't.
> 
> 
> 
> What do they offer but a verbose way to wrap stuff in a class object? How is that any different from wrapping stuff in an object or array literal?

Isolation and grouping of stuff that belongs together. Organizing components and importing them together. It can even be used together with standalone components. You import them into a module and export them and you have a bundle that can be used together.
So for example instead of importing ngIf and ngFor into every component you can import CommonModule.
You could also have patterns were certain components are bundled together with certain implementations of injectables.

**Comment by pfeileon** [comment]
Which part of this list can't be handled by an object or array literal?

```
// shared-modules.ts
export const sharedModules = [
  MySharedStandaloneComponent,
  CommonModule,
  FormsModule,
  RouterModule,
]

// my-standalone.component.ts
@Component({
  template: ``,
  standalone: true,
  imports: [...sharedModules],
})
export class MyStandaloneComponent {}
```

**Comment by hannah23280** [comment]
> Which part of this list can't be handled by an object or array literal?
> 
> ```
> // shared-modules.ts
> export const sharedModules = [
>   MySharedStandaloneComponent,
>   CommonModule,
>   FormsModule,
>   RouterModule,
> ]
> 
> // my-standalone.component.ts
> @Component({
>   template: `<my-shared-standalone />`,
>   standalone: true,
>   imports: [...sharedModules],
> })
> export class MyStandaloneComponent {}
> ```

Maybe the cons are that, we need to manually add the "imports of sharedModules" to **all** the standalone components that need to refer to the sharedModules .

**Comment by pfeileon** [comment]
> The part in which you add a provider.

You either provide in root or directly in the standalone component.

**Comment by DaSchTour** [comment]
> > The part in which you add a provider.
> 
> You either provide in root or directly in the standalone component.

But sometimes you have abstract services or provide something different at the same token.

**Comment by hannah23280** [comment]
```
// shared-modules.ts
export const sharedModules = [
  MySharedStandaloneComponent,
  CommonModule,
  FormsModule,
  RouterModule,
]

// my-standalone.component.ts
@Component({
  template: `<my-shared-standalone />`,
  standalone: true,
  imports: [...sharedModules],
})
export class MyStandaloneComponent {}
```
For this example, does the MySharedStandaloneComponent automatically "inherit" the imports modules from its parent (MyStandalone Component)?

**Comment by twopelu** [comment]
I disagree with the the idea of removing modules, I think it's a good way to organise functionality.
I'd rather revisit the design of services and dependency injection which is one of the major causes of pain in Angular and creates some confusion with how modules work.
In particular, I think that it should work more like in Nest. By default, services in Nest have to be declared and provided only in the module they belong to and then exported in case they are used outside; in particular cases the @Global annotation can be used to make it available everywhere (the docs kindly warn to use this carefully).
This is the opposite to Angular in which by default services are usually provided globally (providedIn: root) and that becomes messy with lazy loaded modules and particular cases.

**Comment by magicmatatjahu** [comment]
@twopelu Hi! COuld you describe why you think that global providers are a problem with lazy loaded modules? When you want to override a provider you declare it in a lazy module, if you don't, you have it available globally. With this approach you can create subdomains (lazy loaded module has another lazy loaded module), which in NestJS are difficult to create because you have to reimport modules to every sub module. I ask out of curiosity, because the child -> parent injector hierarchy is much better in my opinion in Angular than in NestJS (where it doesn't exist), through constant reimporting, which can end up with circulating dependencies in modules.

**Comment by twopelu** [comment]
> @twopelu Hi! Could you describe why you think that global providers are a problem with lazy loaded modules? When you want to override a provider you declare it in a lazy module, if you don't, you have it available globally. With this approach you can create subdomains (lazy loaded module has another lazy loaded module), which in NestJS are difficult to create because you have to reimport modules to every sub module. I ask out of curiosity, because the child -> parent injector hierarchy is much better in my opinion in Angular than in NestJS (where it doesn't exist), through constant reimporting, which can end up with circulating dependencies in modules.

I think is a sum of things:
First of all, I don't like the idea of having different mechanisms for eager and lazy loaded modules, it is simply confusing, so as the methods forRoot and forChild.
Also, It forces the designer of the module to think about how the module is going to be used by others, as it won't be the same if it is loaded eager or lazy. That makes some libraries more complicated. This causes problems with third party libraries and also with self made libraries, as a component may be eager in first place but lately it may have be used lazily, or even both, and then the application might get broken. Probably it is just a design error of the module or library creator, but it is really easy to mess all up in a large codebase with many module dependencies.
To get things worse, there is no way to get to know about the internal behaviour of a library module regarding dependency injection, other than reading the documentation and being really careful, the language / compiler does not enforce anything or warns about it.
TL;DR; The whole thing gets messy and complicated as the application grows, it is error prone.

I have been maintaining a very large codebase with tens of libraries and hundreds of modules and I can assure that it is a pain.
Once in a while, more frequently than I would like to, somebody makes an error an imports a module or provides a service in the wrong way, with causes silent errors in the application. The debugging of this is just horrible, so many time spent following module dependencies and service providers to fix it.
It sounds like a rant (it may be hehe) but I really wish this was easier, I think it could if some design decisions were revisited.

I hope this clarifies my point of view. I really like Angular, it is just some things are overcomplicated.

P.S. The standalone modules in Angular 14 is a good feature that can help in some cases.

**Comment by magicmatatjahu** [comment]
@twopelu Thanks for the answer. I understand your pain. I myself think that the explanation of modules is poorly described in the documentation because it covers simple examples - modules from the community are much more complicated than those in the documentation and people then get lost, I agree with you. On the other hand, I don't think these lazy modules are such a bad thing. They mainly should be used to split applications into smaller ones. Whether they are used well or badly probably depends on the use case - I'm not speaking as an expert here, because the last time I wrote commercially in Angular was about 4 years ago, now I work mainly with React and I miss very much services and dependency injection. Probably if I worked longer with Angular I would see much more problems with current modules. 

As for NestJS, it seems to me that this way implemented in NestJS would not work on the frontend. On the backend it may be a good solution, but on the frontend a parent-child hierarchy is better (React Context and Vue inject functionality also works in this way). On the other hand, it would be nice to have these exports and define which services are private and which are public and to be able to create "lazy" services - globally a module is available, but in lazy modules you define which services are available - if they don't exist create them, if they exist make them available in submodule lazy loaded - because now if you reimport module you exactly create new instance of module (and services "records"), not reusing from parent injector. I don't know if I described it well, but it would be a combination of Angular (parent-child hierarchy) and Nest approach (encapsulation by default, if you want to export something you have to define it as export, and sharing services across lazy loaded modules).

**Comment by grubeninspekteur** [comment]
Deprecation for removal is my main pain point in Angular updates and also other JS libraries. You need to invest dozens of person hours to migrate module-based apps to standalone components, with no business value in return. So I would prefer keeping modules – document that it isn't the preferred way, but remain backwards compatible. Much like the Java Spring framework still works perfectly fine with XML configuration to this day.

**Comment by DaSchTour** [comment]
@twopelu I think methods like forRoot and forChild are there for a reason and just because NgModules are deprecated and removed the reasoning behind them will not go away. So there will be new complicated ways for libraries to solve certain use cases. And I guess that without modules you lose an optional layer that will be replaces with functional components which are extremely ugly and create a lot of clutter in your application.

**Comment by mlc-mlapis** [comment]
@twopelu Where is it defined?
> ... because NgModules are deprecated ...

**Comment by twopelu** [comment]
> @twopelu I think methods like forRoot and forChild are there for a reason and just because NgModules are deprecated and removed the reasoning behind them will not go away. So there will be new complicated ways for libraries to solve certain use cases. And I guess that without modules you lose an optional layer that will be replaces with functional components which are extremely ugly and create a lot of clutter in your application.

and

> @twopelu Where is it defined?
> 
> > ... because NgModules are deprecated ...

Just to clarify, this is what I said first:

> I disagree with the the idea of removing modules, I think it's a good way to organise functionality.

Then I just mentioned some things that may be improved regarding modules and dependency injection, and elaborated a bit more when @magicmatatjahu asked me.

Maybe you are confusing me with the OP :-)

**Comment by TomieAi** [comment]
im too new

> ```
> // shared-modules.ts
> export const sharedModules = [
>   MySharedStandaloneComponent,
>   CommonModule,
>   FormsModule,
>   RouterModule,
> ]
> 
> // my-standalone.component.ts
> @Component({
>   template: `<my-shared-standalone />`,
>   standalone: true,
>   imports: [...sharedModules],
> })
> export class MyStandaloneComponent {}
> ```
> 
> For this example, does the MySharedStandaloneComponent automatically "inherit" the imports modules from its parent (MyStandalone Component)?

is this what really how we import multiple standalone components as a group..  like mimicking ngmodule.  instead of declaring it again and again those tons of imports all over the place.

**Comment by gleb-28** [comment]
You could just not use ngModules at all and all these points would not affect you. There is no reason to take away possibility of using ngModules from other developers

**Comment by thomasgainant-nmt** [comment]
Angular is often chosen over React for its stability and opinionated structure. If you change cornerstones of its structure like NgModules (i.e. changing drastically the opinion behind it), you simply remove the advantage of using such framework.

**Comment by robertIsaac** [comment]
> [@twopelu](https://github.com/twopelu) I think methods like forRoot and forChild are there for a reason and just because NgModules are deprecated and removed the reasoning behind them will not go away. So there will be new complicated ways for libraries to solve certain use cases. And I guess that without modules you lose an optional layer that will be replaces with functional components which are extremely ugly and create a lot of clutter in your application.

@DaSchTour
I disagree, with you, you can use `provideSomething` to do all the things that are possible with forRoot and forChild, ngrx did it, and I have done it with multiple library one of them is open source https://github.com/jfcere/ngx-markdown/pull/484/files#diff-08fc6170344b723d83d255533ee7459a373d15acd5efd78903db80b92c260f45 you can check

I don't see there is any use case that ngModule do better in it, at the opposite it's always easier to do everything without ngModule

**Comment by inmativ** [comment]
Here are two scenarios where it's difficult to do without NgModule:

**One-time initialization on first load**
We need to run some code exactly once during the initial load. The component constructor isn't suitable because it gets called multiple times. APP_INITIALIZER runs at application startup, which isn't what we need either. Wrapping the component in a module becomes a convenient way to organize this code. This approach means:

- I don't have to worry about how my component will be used
- I can be certain the module's constructor will be called exactly once upon first use

**Bundled feature provisioning**
My module provides pipes, components, and directives, plus a shared service that enables:
- Communication between these entities
- Interaction with the host application

Module.forChild() delivers everything with a single line:
- All template entities (pipes/components/directives)
- Their supporting service
- Clean integration point

**Comment by robertIsaac** [comment]
> One-time initialization on first load
We need to run some code exactly once during the initial load. The component constructor isn't suitable because it gets called multiple times. APP_INITIALIZER runs at application startup, which isn't what we need either. Wrapping the component in a module becomes a convenient way to organize this code. This approach means:
I don't have to worry about how my component will be used
I can be certain the module's constructor will be called exactly once upon first use

you can easily do this with using a service provided in root and used only inside this component
here is an example
https://stackblitz.com/edit/stackblitz-starters-argqe6hy?file=src%2Fapp%2Ffoo%2Ffoo.ts

**Comment by robertIsaac** [comment]
> Bundled feature provisioning
My module provides pipes, components, and directives, plus a shared service that enables:
Communication between these entities
Interaction with the host application
Module.forChild() delivers everything with a single line:
All template entities (pipes/components/directives)
Their supporting service
Clean integration point

that's a tradeoff, because if you only need one pipe, you will get the whole module loaded and there is no way to tree shake the others
and from my experience it becomes a dependency hell very easily
because at one point you need to use a component in another component, but another component in the same module needs the first component and it becomes a circular dependency, so you need to go and make another module that has this component and be imported by the two modules, but then you find this component needs another component, and you need to move it as well and you end up moving 10 components
that's why many people started suggesting to use [scam](https://sandroroth.com/blog/angular-shared-scam/) which is standalone minus the module file

**Comment by inmativ** [comment]
> that's a tradeoff, because if you only need one pipe, you will get the whole module loaded and there is no way to tree shake the others and from my experience it becomes a dependency hell very easily because at one point you need to use a component in another component, but another component in the same module needs the first component and it becomes a circular dependency, so you need to go and make another module that has this component and be imported by the two modules, but then you find this component needs another component, and you need to move it as well and you end up moving 10 components that's why many people started suggesting to use [scam](https://sandroroth.com/blog/angular-shared-scam/) which is standalone minus the module file

I agree with you. Standalone components are great. And I would like to fully migrate to them. But the two cases mentioned above are situations where it’s impossible to abandon modules.  

You mention the difficulties that arise due to modules. And that’s a fair point. I try to write components in such a way that they can be easily moved around the codebase without getting tangled with the surrounding code. However, some of my libraries provide a set of components and directives tied together by a service, which simply don’t make sense without each other. They should be imported all at once with a single line and provide the necessary services along with themselves. This can only be achieved with modules.  

The same SCAM allows us to use the module constructor to place code that should run only once. And standalone doesn’t offer any alternatives for this.  

Thus, I stand with you in the belief that we should get rid of modules. But to do this, we need to expand the capabilities of standalone components.  

However, if we do this, how will it differ from modules? It seems we’ll step on the same rake. We need some wise solution.

**Comment by robertIsaac** [comment]
@inmativ can you make an example repo and share it with us
Because I don't understand why the service needs you to import the components and directives 
If it's creating dynamic component and apply directives to them, you only need to import the service and it should just work
If you are passing the component to it, then again you don't need to provide it, just pass it as an argument 
So I can't help you without an example

**Comment by saithis** [comment]
> > that's a tradeoff, because if you only need one pipe, you will get the whole module loaded and there is no way to tree shake the others and from my experience it becomes a dependency hell very easily because at one point you need to use a component in another component, but another component in the same module needs the first component and it becomes a circular dependency, so you need to go and make another module that has this component and be imported by the two modules, but then you find this component needs another component, and you need to move it as well and you end up moving 10 components that's why many people started suggesting to use [scam](https://sandroroth.com/blog/angular-shared-scam/) which is standalone minus the module file
> 
> I agree with you. Standalone components are great. And I would like to fully migrate to them. But the two cases mentioned above are situations where it’s impossible to abandon modules.  
> 
> You mention the difficulties that arise due to modules. And that’s a fair point. I try to write components in such a way that they can be easily moved around the codebase without getting tangled with the surrounding code. However, some of my libraries provide a set of components and directives tied together by a service, which simply don’t make sense without each other. They should be imported all at once with a single line and provide the necessary services along with themselves. This can only be achieved with modules.  
> 
> The same SCAM allows us to use the module constructor to place code that should run only once. And standalone doesn’t offer any alternatives for this.  
> 
> Thus, I stand with you in the belief that we should get rid of modules. But to do this, we need to expand the capabilities of standalone components.  
> 
> However, if we do this, how will it differ from modules? It seems we’ll step on the same rake. We need some wise solution.

You can export an array that contains the standalone components. Then you can either import the standalone components directly or the array to get them all. 

And either use root services for "run only once" code or have a boolean flag to prevent multiple executions.

**Comment by JeanMeche** [comment]
Closing the discussion as nothing is planned today around this. 
Standalone directives are the way to go and we know developers love to use them !
